### PR TITLE
JavaScript: fix the definition of a class

### PR DESCRIPTION
**Pull request by jafl**
Before ES6, there was no way to tell if an individual function was a class constructor.  This is why the convention was set that class names start with uppercase, and function names start with lowercase.

This patch also stops tagging objects as classes.  Objects in JavaScript are just like any other value.  One cannot "new" an object, only a constructor function.

### CODE REVIEW

**Comment by jafl** [comment]
I'm working on fixing the unit tests.

**Comment by jafl** [comment]
I can't figure out why codecov/project is marked fail.

**Comment by masatake** [comment]
In this project, the result reported by codecov is an optional requirement for merging. The coverage is enough high about your change.

**Comment by masatake** [comment]
This change significantly impacts people developing/using IDE that uses ctags internally.
I would like to say this is not a bug fix or enhancement. The change is a kind of "changing the design of parser".
I don't mean I object to the change. However,...
Changing the definition of classes may change how the "tree" at IDE's side bar is rendered.

My JavaScript knowledge is limited, and I don't have much chance to read source code written in JavaScript with ctags.
So I don't have much thought about this change. 
Instead, I would like @jafl to do what he thinks the parser is better; he has already submitted good changes.
I'm always looking for a good parser maintainer knowing the target language.

@b4n, the de-factor parser maintainer of JavaScript, is absent these days. So a person having knowledge, activeness, and motivation should take over the maintainership.  

My only concern is the impact on the other software.
Especially on Geany. Universal Ctags is owing to development efforts at the Geany project.
So I would like to hear the comments from the people on the project.

@jafl, do you use an editor or IDE with ctags?
If yes, how about the impact of this change on the client tool you use?

**Comment by jafl** [comment]
I started thinking about this because my IDE was showing incorrect results, classifying functions and objects as classes.  Since it could have a big impact, I spent days debating whether or not to submit it.  Ultimately, I decided it was the right thing to do because the parser should match the language specification.

It would be great to hear from others if/how this affects their IDE's.

I am also willing to be the maintainer for the JavaScript parser, if that is needed.

File: parsers/jscript.c
Comment (masatake) [inline parsers/jscript.c:1953]: Very very minor item. 
Could you this change from this commit?
Doing this in another commit is acceptable, of course.

File: parsers/jscript.c
Comment (masatake) [inline parsers/jscript.c:432]: Please, access the internal of vString directly only when you have a strong reason.
Instead, let's use vStringValue like:
```
vStringValue (name->string);
```

File: parsers/jscript.c
Comment (masatake) [inline parsers/jscript.c:434]: Let's use vStringValue.
If you find lines that access `->buffer` directly in jscript.c (and the other files), they can be candidates for fixing in another pull request:-)

File: parsers/jscript.c
Comment (jafl) [reply parsers/jscript.c:432]: Thanks!  That's much better.

File: parsers/jscript.c
Comment (masatake) [inline parsers/jscript.c:2336]: With the change, the value returned from parserMethods are ignored, and the tag for the anonymous object is not made. I think this is not a good idea. Could you keep the original code with just changing the kind like:
```c
			has_methods = parseMethods(token, name, false);
			if (has_methods)
				makeJsTagCommon (name, JSTAG_VARIABLE, NULL, NULL, anonObject);
			else
			{
```
?

File: parsers/jscript.c
Comment (masatake) [inline parsers/jscript.c:1943]: To extract an object having a method or a property as an anonymous tag, we must keep this local variable. See below.

**Comment by masatake** [comment]
@techee, @eht16, and the other Geany developers.
I would like to merge this pull request.
However, this change may have a big impact on Geany.
Could you give us comments?

**Comment by masatake** [comment]
I wonder we need a mechanism for declaring the version of a parser like:

```C
extern parserDefinition* JavaScriptParser (void)
{
	// .jsx files are JSX: https://facebook.github.io/jsx/
	// which have JS function definitions, so we just use the JS parser
	static const char *const extensions [] = { "js", "jsx", "mjs", NULL };
	static const char *const aliases [] = { "js", "node", "nodejs",
	                                        "seed", "gjs",
											/* Used in PostgreSQL
											 * https://github.com/plv8/plv8 */
											"v8",
											NULL };
	parserDefinition *const def = parserNew ("JavaScript");
        def->version = {[CURRENT] = 1, [REVISION] = 0, [AGE] = 0}; // as describe in info page of libtool
...
```

The parser version information may be emitted like:
```
$ cat tags
...
!_TAG_PARSER_VERSION!JavaScript	current	/1/
!_TAG_PARSER_VERSION!JavaScript	revision	/0/
!_TAG_PARSER_VERSION!JavaScript	age	/0/
...
```

Geany includes ctags code. So Geany doesn't need these pseudo tags about version information .
Other client tools reading tags file can utilize the pseudo tags.

**Comment by jafl** [comment]
Should the parser version be included in this PR?

**Comment by masatake** [comment]
> Should the parser version be included in this PR?

No. Sorry for making you confused. It is just an idea.

**Comment by jafl** [comment]
I will merge the updates from #3357 into this PR after that PR is merged.

File: Units/parser-javascript.r/js-empty-class-name.d/expected.tags
Comment (masatake) [inline Units/parser-javascript.r/js-empty-class-name.d/expected.tags:1]: The value for scope field is `class:anonymousObject4ca5b60a0105`.
However, `anonymousObject4ca5b60a0105` is tagged as a variable.
So I think the value of the field should be:
```
variable:anonymousObject4ca5b60a0105
```

File: Units/parser-javascript.r/github-issue-780.d/expected.tags
Comment (masatake) [inline Units/parser-javascript.r/github-issue-780.d/expected.tags:3]: Though `anonymousObject6b70cce50105` is tagged as a variable, it is referred as a class in the tag of `jump`.

**Comment by techee** [comment]
@masatake Is this just a matter of renaming anonymous tags and reporting some with a different kind? A don't see any problem with that, we'll just have to update our kind mappings but we are prepared for such situations.

File: Units/parser-javascript.r/js-empty-class-name.d/expected.tags
Comment (jafl) [reply Units/parser-javascript.r/js-empty-class-name.d/expected.tags:1]: Good point.  This requires that I tackle a FIXME in the code.  The patch is going to get bigger :)

**Comment by masatake** [comment]
> @masatake Is this just a matter of renaming anonymous tags and reporting some with a different kind? A don't see any problem with that, we'll just have to update our kind mappings but we are prepared for such situations.

As far as the source code area I reviewed,  we don't have to rename the anonymous tags in this case.
Even if we have to rename them, it is easy.

We have to update ` extensionFields.scopeKindIndex` as you already know.

After reading your code, updating the `socpeKindIndex` is more complicated than I assumed.

I want to carefully review your new change (difference between  1bab2a2 and 5acd011). 
Could you do "push --force" to replace the current single commit with the original and new changes?
`git reflog` may help you.

**Comment by masatake** [comment]
@techee 

> @masatake Is this just a matter of renaming anonymous tags and reporting some with a different kind? A don't see any problem with that, we'll just have to update our kind mappings but we are prepared for such situations.

The name for anonymous tags may not be an issue here.


In the original code, methods and properties have a class in their scope fields:

input.js:
```
var o = {
    init: function () {
    },
    fini: function () {
    },
};
var p = 1
```

tags output with the original code:
```
init	foo.js	/^    init: function () {$/;"	kind:method	scope:class:o	signature:()	roles:def
fini	foo.js	/^    fini: function () {$/;"	kind:method	scope:class:o	signature:()	roles:def
o	foo.js	/^var o = {$/;"	kind:class	roles:def
p	foo.js	/^var p = 1$/;"	kind:variable	roles:def
```

tags output with this change:
```
init	foo.js	/^    init: function () {$/;"	kind:method	scope:variable:o	signature:()	roles:def
fini	foo.js	/^    fini: function () {$/;"	kind:method	scope:variable:o	signature:()	roles:def
o	foo.js	/^var o = {$/;"	kind:variable	lroles:def
p	foo.js	/^var p = 1$/;"	kind:variable	roles:def
```

If Geany expects methods (and properties) may have "class" as their scopes, this pull request breaks the expectation.
With this request, the tags output lost some information. With the original code, `p` and `q` have different kinds. With this pull request, `p` and `q` have the same kind, `variable`. I wonder if this is a matter or not in Geany.

**Comment by masatake** [comment]
@jafl 

After reading your code, updating the `socpeKindIndex` is more complicated than I assumed.

I want to carefully review your new change (difference between  1bab2a2 and 5acd011). 
Could you do "push --force" to replace the current single commit with the original and new changes?
`git reflog` may help you.

**Comment by masatake** [comment]
```
var o = {
    init: function () {
    },
    fini: function () {
    },
};
var p = 1
```

In my understanding, What @jafl wants to say is `o` is not a `class`.
This is understandable even with my limited knowledge of JavaScript.

I will talk about the next steps AFTER merging this pull request.
In lexically, `o` is a variable. However, I think people using Javascript may want to say `o` is an object.
In the other words, they expect the following tags output:
```
init	foo.js	/^    init: function () {$/;"	kind:method	scope:object:o	signature:()	roles:def
fini	foo.js	/^    fini: function () {$/;"	kind:method	scope:object:o	signature:()	roles:def
o	foo.js	/^var o = {$/;"	kind:object	roles:def
p	foo.js	/^var p = 1$/;"	kind:variable	roles:def
```

I'm comfortable with `o` and `p` being tagged differently.
 
The algorithm emitting `o` is:
0. introduce a new kind "o/object".
1. make a tag for `o` with `variable` kind as this pull request does.
2. count the methods and properties defined in `o` => 2.
3. change the kind of `o` to `object` if the counter > 0.
4. update the scope fields of tags having `o` as their scope.

The following case doesn't work:
```
var gen = function () {
 return {
    init: function () {
    },
    fini: function () {
    }
  };
}
var o = get()
var p = 1
```

In this case `o` is tagged as a variable. 

How about an array? Introducing the "array" kind?
I'm not sure:-)

**Comment by techee** [comment]
> If Geany expects methods (and properties) may have "class" as their scopes, this pull request breaks the expectation.
With this request, the tags output lost some information. With the original code, p and q have different kinds. With this pull request, p and q have the same kind, variable. I wonder if this is a matter or not in Geany.

@masatake I don't think this really matters to Geany, it should be no problem for us.

**Comment by masatake** [comment]
@jafl, thank you.

**Comment by masatake** [comment]
Now I understood your change. 
Introducing corkAPI may be better. But the way you took is acceptable.

I will review  5acd011 directly. We don't need 1bab2a2 anymore.

File: parsers/jscript.c
Comment (masatake) [inline parsers/jscript.c:2116]: These two lines imply it is better to introduce a helper function named `clearTokenScope`.
```c
static void clearTokenScope(tokenInfo *token)
{
   vStringClear (token->scope);
   token->scopeKinde = KIND_GHOST_INDEX;
}
```

File: parsers/jscript.c
Comment (masatake) [inline parsers/jscript.c:1930]: How about introducing something like `fillTokenScope(tokenInfo *token, vString *saveScope, int saveScopeKind);`?

Comment (masatake) [reply parsers/jscript.c:1930]: `setScope` may be better as the name of the function.

File: parsers/jscript.c
Comment (masatake) [reply parsers/jscript.c:2116]: `clearScope` or `resetScope` may be better.

File: Units/review-needed.r/simple.js.t/expected.tags
Comment (masatake) [inline Units/review-needed.r/simple.js.t/expected.tags:2]: Very interesting.
I have many comments about this line.
...

**Comment by masatake** [comment]
When, I understand the reason why you introduce the "scope" kind, I wonder what we should do.

This pull request will become much larger. The goal of change becomes ambiguous.
... I think introducing #2115 first is better. This pull request, switching class to variable kinds, may come later.

I would like to try to rebase #2115 to the latest master. So I would like you to review my approach.
Could you wait for a week or so?

In addition to corkAPI, I would like you to know the concepts, roles and reference tags

See http://docs.ctags.io/en/latest/man/ctags-lang-python.7.html and http://docs.ctags.io/en/latest/man/ctags.1.html?highlight=roles#roles .

With combining these concepts, we can fill the scope fields of tags cleanly.

File: Units/review-needed.r/simple.js.t/expected.tags
Comment (jafl) [reply Units/review-needed.r/simple.js.t/expected.tags:2]: Yes, this is a tricky issue.  The parser doesn't always know what the thing is.  Perhaps the cork queue will help solve this?

**Comment by jafl** [comment]
NetBSD unit tests failed due to "Transient resolver failure"

**Comment by masatake** [comment]
@jafl, unexpected failures on *BSD happen frequently. In such cases, could you press "Re-run jobs" by yourself? The question is whether you can do it with your account.

**Comment by masatake** [comment]
Oh, I already did "Re-run jobs" though I wanted you to try.

**Comment by jafl** [comment]
I do not see a "Re-run jobs" button.

**Comment by masatake** [comment]
> I do not see a "Re-run jobs" button.

I see. I'm looking for a way to make it visible to you.

**Comment by masatake** [comment]
Like #3424, I guess the change for [Units/parser-javascript.r/1880687.js.d/input.js](https://github.com/universal-ctags/ctags/pull/3363/files#diff-f31b95120b6ea91dcf46b0b2d091314d2ef1cb8bbb74a85c3ece5a63e166f675) in this pull request is not intentional. Could you rebase this pull request on the latest master branch?

File: parsers/jscript.c
Comment (masatake) [inline parsers/jscript.c:604]: In my understanding, a cast operator `(unsigned int)` is needed in the front of `*p`.
isupper() expects int but only -1 is the acceptable negative value; the behavior of isupper() is not defined in the standard C spec.
On the other hand, we cannot have an assume whether char is `unsigned char` or `singed char`. So `*p` can be a negative value other than -1.

See https://github.com/universal-ctags/ctags/pull/1697 .

Comment (jafl) [reply parsers/jscript.c]: Good point.  Fixed.

**Comment by jafl** [comment]
@masatake You will have to re-run the BSD tests.

**Review by masatake — APPROVED** [review]

**Comment by masatake** [comment]
If one in the project becomes absent, the God (, something you believe, or something supernatural) dispatches another hacker to this project. It is known as the "law of conservation of the number of ctags developers".

@jafl, please, merge this!

**Comment by jafl** [comment]
I don't see a button for merging.

**Comment by masatake** [comment]
Thank you.

File: Units/parser-javascript.r/1795612.js.d/expected.tags
Comment (b4n) [inline Units/parser-javascript.r/1795612.js.d/expected.tags:1]: Is this change expected?  Admittedly `Test` is not actually declared in this file so creating a tag for it is a bit artificial, but this change looks a bit unexpected (it's at least not obvious to me which change did this, so I can't tell if it was intentional).

I don't really mind either way as there's no definite answer (it's not even proper JS given `Test` is not even imported or anything), but I'm wondering if this was on purpose or an accident.

Comment (masatake) [reply Units/parser-javascript.r/1795612.js.d/expected.tags:1]: This looks like a bug about the granularity of tagging.

Looking at this again, a tag (`Test.RPC`) with more than one token (`Test` and `RPC`) looks strange. (as far as comparing with 
Units/parser-javascript.r/js-class-related-unterminated.d/expected.tags)

My expectation (--extras=+{reference} --fields=+{roles}{scope}):

```
RPC	input.js	/^Test.RPC =$/;"	p	scope:variable:Test
Test	input.js	/^Test.RPC =$/;"	v	roles:chainElt
```

I introduced the `chainElt` roles for avoiding more-than-one-token tag entries.


 
There is room for discussion about the kinds (p for `RPC` and v for `Test`), but there will be no perfect answer. It is up to the parser maintainer.

Comment (masatake) [reply Units/parser-javascript.r/1795612.js.d/expected.tags:1]: I'll convert this to an issue.

File: Units/parser-javascript.r/bug1950327.js.d/expected.tags
Comment (masatake) [inline Units/parser-javascript.r/bug1950327.js.d/expected.tags:4]: This is just another two-tokens tag.
I think this should be:
```
dirTab	input.js	/^		container.dirtyTab = {'url': false, 'title':false, 'snapshot':false, '*': false}		$/;"	p	variable:container
container	input.js	/^		container.dirtyTab = {'url': false, 'title':false, 'snapshot':false, '*': false}		$/;"	v	roles:chainElt
```

What the scope for `cotainer` is is another issue.

File: Units/parser-javascript.r/1795612.js.d/expected.tags
Comment (masatake) [reply Units/parser-javascript.r/1795612.js.d/expected.tags:1]:
```diff
diff --git a/parsers/jscript.c b/parsers/jscript.c
index 3788cc090..f54c432da 100644
--- a/parsers/jscript.c
+++ b/parsers/jscript.c
@@ -485,7 +485,7 @@ static int makeJsTagCommon (const tokenInfo *const token, const jsKind kind,
 
 	const char *p;
 	char *name_chain = NULL;
-	if (!token->dynamicProp && kind != JSTAG_PROPERTY &&  (p = strrchr (name, '.')) != NULL )
+	if (!token->dynamicProp && (p = strrchr (name, '.')) != NULL )
 	{
 		if ((p - name) != 0)
 			name_chain = eStrndup (name, (size_t) (p - name));
```

With this, we can avoid the two-tokens tags.
It is easy to fix the parser as we want it, but it is not easy to decide how the parser should be.

Comment (masatake) [reply Units/parser-javascript.r/1795612.js.d/expected.tags:1]: We should open two issues. I will open issues for them.

**Pull request merged**
